src/P3_particle_properties.jl

```diff
@@ -1,333 +1,498 @@
-    return 6 * α_va * (D_cr^β_m2 - D_gr^β_m2) / FT(π) / β_m2 /
-           max(D_cr - D_gr, eps(FT))
-end
+get_D_th(mass::MassPowerLaw, ρ_i) = get_threshold(mass, ρ_i)
 
 """
-    thresholds(p3, ρ_r, F_rim)
+    get_D_gr(mass::MassPowerLaw, ρ_g)
 
- - p3 - a struct with P3 scheme parameters
- - ρ_r - rime density (L_rim/B_rim) [kg/m^3]
- - F_rim - rime mass fraction (L_rim / L_ice) [-]
+Return the size of equal mass for graupel and unrimed ice [meters]
 
-Solves the nonlinear system consisting of D_cr, D_gr, ρ_g, ρ_d
-for a given rime density and rime mass fraction.
-Returns a named tuple containing:
- - D_cr - is the threshold size separating partially rimed ice and graupel [m],
- - D_gr - is the threshold size separating graupel and dense nonspherical ice [m],
- - ρ_g - is the effective density of a spherical graupel particle [kg/m3],
- - ρ_d - is the density of the unrimed portion of the particle [kg/m3],
+See Eq. 15 in [MorrisonMilbrandt2015](@cite).
 """
-function thresholds(p3::PSP3{FT}, ρ_r::FT, F_rim::FT) where {FT}
+get_D_gr(mass::MassPowerLaw, ρ_g) = get_threshold(mass, ρ_g)
 
-    @assert F_rim >= FT(0)   # rime mass fraction must be positive ...
-    @assert F_rim < FT(1)    # ... and there must always be some unrimed part
+"""
+    get_D_cr(mass::MassPowerLaw, ρ_g, F_rim)
 
-    if F_rim == FT(0)
-        return (; D_cr = FT(0), D_gr = FT(0), ρ_g = FT(0), ρ_d = FT(0))
-    else
-        @assert ρ_r > FT(0)   # rime density must be positive ...
-        @assert ρ_r <= p3.ρ_l # ... and as a bulk ice density can't exceed the density of water
-
-        P3_problem(ρ_d) =
-            ρ_d - ρ_d_helper(
-                p3,
-                D_cr_helper(p3, F_rim, ρ_g_helper(ρ_r, F_rim, ρ_d)),
-                D_gr_helper(p3, ρ_g_helper(ρ_r, F_rim, ρ_d)),
-            )
-
-        ρ_d =
-            RS.find_zero(
-                P3_problem,
-                RS.SecantMethod(FT(0), FT(1000)),
-                RS.CompactSolution(),
-            ).root
-        ρ_g = ρ_g_helper(ρ_r, F_rim, ρ_d)
-
-        return (;
-            D_cr = D_cr_helper(p3, F_rim, ρ_g),
-            D_gr = D_gr_helper(p3, ρ_g),
-            ρ_g,
-            ρ_d,
-        )
-    end
-end
+Return the size of equal mass for graupel and partially rimed ice [meters]
 
+See Eq. 14 in [MorrisonMilbrandt2015](@cite).
 """
-    p3_F_liq_average(F_liq, X_ice, X_liq)
+get_D_cr(mass::MassPowerLaw, ρ_g, F_rim) =
+    get_threshold(mass, ρ_g * (1 - F_rim))
 
- - F_liq - liquid fraction (L_liq / L_p3_tot)
- - X_ice - ice core parameterization (i.e. mass, etc)
- - X_liq - liquid part parameterization
+"""
+    weighted_average(f_a, a, b)
 
-Returns the liquid fraction weighted average of X_ice and X_liq.
+Return the weighted average of `a` and `b` with fraction `f_a`,
+    
+```math
+f_a ⋅ a + (1 - f_a) ⋅ b
+```
 """
-function p3_F_liq_average(F_liq::FT, X_ice::FT, X_liq::FT) where {FT}
-    return (1 - F_liq) * X_ice + F_liq * X_liq
+function weighted_average(f_a, a, b)
```

Comment:
Looks like it's only used once, so maybe we don't need to define a special function for it?

Reply:
this will be needed for liquid fraction calculations.